engine/execution/storehouse/block_end_snapshot.go

```diff
@@ -0,0 +1,91 @@
+package storehouse
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/onflow/flow-go/engine/execution"
+	"github.com/onflow/flow-go/fvm/storage/snapshot"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+var _ snapshot.StorageSnapshot = (*BlockEndStateSnapshot)(nil)
+
+// BlockEndStateSnapshot represents the storage at the end of a block.
+type BlockEndStateSnapshot struct {
+	storage execution.RegisterStore
+
+	blockID flow.Identifier
+	height  uint64
+
+	mutex     sync.RWMutex
+	readCache map[flow.RegisterID]flow.RegisterValue // cache the reads from storage at baseBlock
+}
+
+// the caller must ensure the block height is for the given block
+func NewBlockEndStateSnapshot(
+	storage execution.RegisterStore,
+	blockID flow.Identifier,
+	height uint64,
+) *BlockEndStateSnapshot {
+	return &BlockEndStateSnapshot{
+		storage:   storage,
+		blockID:   blockID,
+		height:    height,
+		readCache: make(map[flow.RegisterID]flow.RegisterValue),
+	}
+}
+
+// Get returns the value of the register with the given register ID.
+// It returns:
+// - (value, nil) if the register exists
+// - (nil, nil) if the register does not exist
+// - (nil, storage.ErrHeightNotIndexed) if the height is below the first height that is indexed.
+// - (nil, storehouse.ErrNotExecuted) if the block is not executed yet
+// - (nil, storehouse.ErrNotExecuted) if the block is conflicting with finalized block
+// - (nil, err) for any other exceptions
+func (s *BlockEndStateSnapshot) Get(id flow.RegisterID) (flow.RegisterValue, error) {
+	value, ok := s.getFromCache(id)
+	if ok {
+		if value == nil {
+			return nil, storage.ErrNotFound
+		}
+		return value, nil
```

Comment:
Shouldn't this just return (value, nil) now. If value is nil we want to return nil, right?

Reply:
yup, updated